sealing/garbage.go

```diff
@@ -33,97 +31,128 @@ func (m *Sealing) pledgeReader(size abi.UnpaddedPieceSize, parts uint64) io.Read
+			Piece:    existing[idx].Piece,
+			DealInfo: &existing[idx].DealInfo,
+		}
+	}
+
+	schedule, err := m.selfDealPolicy.Schedule(ctx, append(fillers, existingPrime...)...)
+	if err != nil {
+		return nil, handle("failed to create self-deal schedule", err)
+	}
+
+	// we send self-deals only for the filler pieces, as the other pieces are
+	// already associated with on-chain deals
+	fillersPrime := make([]abi.PieceInfo, len(fillers))
+	for idx := range fillersPrime {
+		fillersPrime[idx] = abi.PieceInfo{
+			Size:     fillers[idx].Piece.Size,
+			PieceCID: fillers[idx].Piece.PieceCID,
+		}
+	}
+
+	mcid, err := m.api.SendSelfDeals(ctx, schedule.StartEpoch, schedule.EndEpoch, fillersPrime...)
 	if err != nil {
-		return nil, err
+		return nil, handle("failed to send self-deals to node", err)
 	}
 
 	dealIDs, exitCode, err := m.api.WaitForSelfDeals(ctx, mcid)
 	if err != nil {
-		return nil, err
+		return nil, handle("node produced an error waiting for self deals", err)
 	}
 
 	if exitCode != 0 {
-		err := xerrors.Errorf("publishing deal failed: exit %d", exitCode)
-		log.Error(err)
-		return nil, err
+		return nil, handle("publishing deal failed: exit %d", exitCode)
 	}
 
-	if len(dealIDs) != len(sizes) {
-		err := xerrors.New("got unexpected number of DealIDs from PublishStorageDeals")
-		log.Error(err)
-		return nil, err
+	if len(dealIDs) != len(fillerPieceSizes) {
+		return nil, handle("got unexpected number of deal IDs from PublishStorageDeals (len(dealIDs)=%d != len(fillerPieceSizes)=%d)", len(dealIDs), len(fillerPieceSizes))
 	}
 
-	out := make([]node.Piece, len(sizes))
-	for i, size := range sizes {
-		ppi, err := m.sb.AddPiece(ctx, size, sectorNum, m.pledgeReader(size, uint64(runtime.NumCPU())), existingPieceSizes)
-		if err != nil {
-			return nil, xerrors.Errorf("add piece: %w", err)
-		}
+	// the sizes of the pieces already written to the sector into which the
+	// junk pieces will be written
+	existingSizes := make([]abi.UnpaddedPieceSize, len(existing))
+	for idx := range existingSizes {
+		existingSizes[idx] = existing[idx].Piece.Size.Unpadded()
+	}
 
-		existingPieceSizes = append(existingPieceSizes, size)
+	var out []node.PieceWithDealInfo
 
-		out[i] = node.Piece{
-			DealID:   dealIDs[i],
-			Size:     ppi.Size.Padded(),
-			PieceCID: commcid.PieceCommitmentV1ToCID(ppi.CommP[:]),
+	for idx := range fillerPieceSizes {
+		ppi, err := m.sb.AddPiece(ctx, fillerPieceSizes[idx], sectorNum, m.pledgeReader(fillerPieceSizes[idx], uint64(runtime.NumCPU())), existingSizes)
+		if err != nil {
+			return nil, handle("add piece: %w", err)
 		}
+
+		existingSizes = append(existingSizes, fillerPieceSizes[idx])
+
+		out = append(out, node.PieceWithDealInfo{
+			Piece: abi.PieceInfo{
+				Size:     ppi.Size.Padded(),
+				PieceCID: commcid.PieceCommitmentV1ToCID(ppi.CommP[:]),
+			},
+			DealInfo: node.DealInfo{
+				DealID:       dealIDs[idx],
+				DealSchedule: schedule,
+			},
+		})
 	}
 
+	out = append(out, existing...)
```

Comment:
Not sure if this matters, but the ordering of the pieces returned here (and then written to the state machine maybe?) is reversed.. `[fillers..., existing...]`

Reply:
A fine point!